File: api/models/evidence_packing.py
"""Closed operational retrieval plans; grouping never proves independence."""
from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

RoleHint = Literal["methods", "results", "table", "other"]
GroupBasis = Literal["accepted_work_mapping", "source_snapshot", "legacy_paper"]
ExclusionReason = Literal[
    "payload_budget", "base_payload_budget", "chunk_limit", "source_limit", "work_limit",
    "duplicate_content", "role_already_represented", "not_complementary_original",
]
SummaryReason = Literal[
    "packing_not_requested", "no_admitted_candidates", "base_payload_budget_exceeded", "payload_budget_excluded",
    "selection_limits_applied", "duplicate_content_removed", "complementarity_not_established",
    "packing_unavailable", "selected_context_withheld",
]


class _Closed(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, frozen=True, allow_inf_nan=False)


class PackingCandidate(_Closed):
    """Private, already-admitted identities supplied by a trusted DB resolver.

    No permission/approval status is accepted from an untrusted request body.
    Accepted Work IDs are diversity hints, not independent scientific roots.
    """
    chunk_id: str = Field(min_length=1, max_length=200)
    paper_id: str = Field(min_length=1, max_length=100)
    source_snapshot_sha256: str | None = Field(default=None, pattern=r"^[0-9a-f]{64}$")
    accepted_work_id: str | None = None
    content_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    chunk_kind: Literal["original_passage", "abstract", "derived_fact", "legacy_unknown"]
    role_hint: RoleHint = "other"

    @field_validator("chunk_id", "paper_id")
    @classmethod
    def bounded_identifier(cls, value):
        if not value.strip() or any(ord(character) < 32 or ord(character) == 127 for character in value):
            raise ValueError("A bounded nonblank identifier is required")
        return value

    @field_validator("accepted_work_id")
    @classmethod
    def canonical_work(cls, value):
        if value is not None and str(UUID(value)) != value:
            raise ValueError("Accepted Work diversity requires its canonical UUID")
        return value


class EvidencePackingSelection(_Closed):
    version: Literal["evidence-pack-item/1.0.0"] = "evidence-pack-item/1.0.0"
    position: int = Field(ge=1, le=20)
    chunk_id: str = Field(min_length=1, max_length=200)
    source_group_id: str = Field(pattern=r"^src:[0-9a-f]{64}$")
    source_snapshot_sha256: str | None = Field(default=None, pattern=r"^[0-9a-f]{64}$")
    diversity_group_id: str = Field(pattern=r"^div:[0-9a-f]{64}$")
    source_group_basis: Literal["source_snapshot", "legacy_paper"]
    group_basis: GroupBasis
    role_hint: RoleHint
    selection_reason: Literal["source_diversity", "source_coverage", "complementary_role"]
    scientific_acceptance: Literal[False] = False

    @field_validator("scientific_acceptance", mode="before")
    @classmethod
    def no_acceptance(cls, value):
        if value is not False:
            raise ValueError("A packing decision never grants scientific acceptance")
        return value

    @model_validator(mode="after")
    def consistent_basis(self):
        if (self.source_group_basis == "source_snapshot") != (self.source_snapshot_sha256 is not None):
            raise ValueError("Source grouping must disclose whether an exact snapshot was bound")
        if self.group_basis != "accepted_work_mapping" and self.group_basis != self.source_group_basis:
            raise ValueError("Unmapped diversity must retain its exact source grouping basis")
        if self.source_group_basis == "legacy_paper" and self.selection_reason == "complementary_role":
            raise ValueError("Unresolved legacy source snapshots cannot be complemented")
        if self.selection_reason == "complementary_role" and self.role_hint == "other":
            raise ValueError("Only explicit retrieval role hints provide complementarity")
        return self


class EvidencePackingExclusion(_Closed):
    chunk_id: str = Field(min_length=1, max_length=200)
    reason_code: ExclusionReason


class EvidencePackingPlan(_Closed):
    version: Literal["evidence-packing/1.0.0"] = "evidence-packing/1.0.0"
    status: Literal["packed", "empty", "base_budget_exceeded"]
    selected: list[EvidencePackingSelection] = Field(default_factory=list, max_length=20)
    excluded: list[EvidencePackingExclusion] = Field(default_factory=list, max_length=300)
    candidate_count: int = Field(ge=0, le=300)
    selected_count: int = Field(ge=0, le=20)
    source_group_count: int = Field(ge=0, le=20)
    diversity_group_count: int = Field(ge=0, le=20)
    payload_bytes: int = Field(ge=0, le=64 * 1024 * 1024)
    byte_budget: int = Field(ge=1, le=16 * 1024 * 1024)
    byte_count_method: Literal["utf8-full-payload/1"] = "utf8-full-payload/1"
    max_chunks: int = Field(ge=1, le=20)
    max_per_source: int = Field(ge=1, le=3)
    max_per_work: int = Field(ge=1, le=3)
    independent_support_count: None = None
    independence_status: Literal["independence_not_established"] = "independence_not_established"
    scientific_acceptance: Literal[False] = False

    @field_validator("scientific_acceptance", mode="before")
    @classmethod
    def no_acceptance(cls, value):
        if value is not False:
            raise ValueError("Grouping is not scientific acceptance")
        return value

    @model_validator(mode="after")
    def coherent_plan(self):
        identifiers = [item.chunk_id for item in [*self.selected, *self.excluded]]
        if len(set(identifiers)) != len(identifiers) or len(identifiers) != self.candidate_count:
            raise ValueError("Every admitted candidate must have exactly one packing disposition")
        if self.selected_count != len(self.selected) or self.selected_count > self.max_chunks:
            raise ValueError("Packing selection count mismatch")
        if [item.position for item in self.selected] != list(range(1, self.selected_count + 1)):
            raise ValueError("Citation positions must be contiguous and ordered")
        if (self.source_group_count != len({item.source_group_id for item in self.selected})
                or self.diversity_group_count != len({item.diversity_group_id for item in self.selected})):
            raise ValueError("Operational grouping count mismatch")
        seen_sources, seen_diversity, roles = set(), set(), {}
        for item in self.selected:
            expected = ("source_diversity" if item.diversity_group_id not in seen_diversity else
                        "source_coverage" if item.source_group_id not in seen_sources else "complementary_role")
            if item.selection_reason != expected:
                raise ValueError("Packing selection reason must match its ordered grouping")
            if item.selection_reason == "complementary_role" and item.role_hint in roles.get(item.source_group_id, set()):
                raise ValueError("A repeated role is not complementary source coverage")
            source_members = [row for row in self.selected if row.source_group_id == item.source_group_id]
            if len(source_members) > (1 if item.source_group_basis == "legacy_paper" else self.max_per_source):
                raise ValueError("Source group packing limit exceeded")
            if len({(row.source_group_basis, row.diversity_group_id, row.group_basis) for row in source_members}) != 1:
                raise ValueError("One exact source group cannot have conflicting diversity bindings")
            if len({row.source_snapshot_sha256 for row in source_members}) != 1:
                raise ValueError("One exact source group cannot mix source snapshots")
            if item.group_basis == "accepted_work_mapping" and sum(
                    row.diversity_group_id == item.diversity_group_id for row in self.selected) > self.max_per_work:
                raise ValueError("Accepted Work diversity limit exceeded")
            seen_sources.add(item.source_group_id)
            seen_diversity.add(item.diversity_group_id)
            roles.setdefault(item.source_group_id, set()).add(item.role_hint)
        if self.status == "packed":
            valid = bool(self.selected) and self.payload_bytes <= self.byte_budget
        elif self.status == "empty":
            valid = not self.selected and self.payload_bytes <= self.byte_budget
        else:
            valid = not self.selected and self.payload_bytes > self.byte_budget
        if not valid:
            raise ValueError("Packing state and complete-payload byte accounting disagree")
        return self


class EvidencePackingSummary(_Closed):
    """Public accounting only: no withheld candidate IDs or raw Work UUIDs."""
    version: Literal["evidence-packing/1.0.0"] = "evidence-packing/1.0.0"
    status: Literal["not_requested", "packed", "empty", "base_budget_exceeded", "withheld", "unavailable"] = "not_requested"
    candidate_count: int = Field(default=0, ge=0, le=300)
    selected_count: int = Field(default=0, ge=0, le=20)
    source_group_count: int = Field(default=0, ge=0, le=20)
    diversity_group_count: int = Field(default=0, ge=0, le=20)
    payload_bytes: int | None = Field(default=None, ge=0, le=64 * 1024 * 1024)
    byte_budget: int | None = Field(default=None, ge=1, le=16 * 1024 * 1024)
    byte_count_method: Literal["utf8-full-payload/1"] = "utf8-full-payload/1"
    max_chunks: int = Field(default=20, ge=1, le=20)
    max_per_source: int = Field(default=3, ge=1, le=3)
    max_per_work: int = Field(default=3, ge=1, le=3)
    reason_counts: dict[ExclusionReason, int] = Field(default_factory=dict)
    reason_codes: list[SummaryReason] = Field(default_factory=list, max_length=9)
    independent_support_count: None = None
    independence_status: Literal["independence_not_established"] = "independence_not_established"
    scientific_acceptance: Literal[False] = False

    @field_validator("scientific_acceptance", mode="before")
    @classmethod
    def no_acceptance(cls, value):
        if value is not False:
            raise ValueError("Grouping is not scientific acceptance")
        return value

    @field_validator("reason_counts")
    @classmethod
    def positive_counts(cls, value):
        if any(type(count) is not int or not 1 <= count <= 300 for count in value.values()):
            raise ValueError("Packing exclusion counts must be bounded positive integers")
        return value

    @model_validator(mode="after")
    def coherent_summary(self):
        if (not 0 <= self.diversity_group_count <= self.source_group_count <= self.selected_count <= self.candidate_count
                or self.selected_count > self.max_chunks or len(set(self.reason_codes)) != len(self.reason_codes)):
            raise ValueError("Inconsistent public packing counts")
        if self.status in {"not_requested", "withheld", "unavailable"}:
            if self.selected_count or self.source_group_count or self.diversity_group_count or self.payload_bytes is not None or self.reason_counts:
                raise ValueError("Unavailable or withheld plans cannot disclose a previous selected context")
            if self.status == "not_requested" and (self.candidate_count or self.byte_budget is not None):
                raise ValueError("Unrequested packing has no measured inventory or budget")
        else:
            if self.payload_bytes is None or self.byte_budget is None:
                raise ValueError("A completed plan needs exact complete-payload accounting")
            if sum(self.reason_counts.values()) != self.candidate_count - self.selected_count:
                raise ValueError("Public exclusion counts must account for the complete inventory")
            if self.status == "base_budget_exceeded":
                valid = self.selected_count == 0 and self.payload_bytes > self.byte_budget
            else:
                valid = self.payload_bytes <= self.byte_budget and (self.selected_count > 0) == (self.status == "packed")
            if not valid:
                raise ValueError("Public packing state disagrees with its byte budget")
        return self

File: frontend/lib/evidence-packing.ts
/** Operational display guards, never evidence/root/permission adjudication. */
import type { EvidencePackingSelection, EvidencePackingSummary, RagInputBudgetReport } from "@/lib/api";
import { knownEvidenceProvenance } from "@/lib/evidence-provenance";

type Row = Record<string, unknown>;
const row = (value: unknown): value is Row => value !== null && typeof value === "object" && !Array.isArray(value);
function keys(value: unknown, expected: string[]): value is Row {
  return row(value) && Object.keys(value).length === expected.length && expected.every(key => Object.hasOwn(value, key));
}
const oneOf = (value: unknown, choices: string[]) => typeof value === "string" && choices.includes(value);
const integer = (value: unknown, min: number, max: number): value is number => typeof value === "number" && Number.isSafeInteger(value) && value >= min && value <= max;
const nullableInteger = (value: unknown, min: number, max: number) => value === null || integer(value, min, max);
const hash = (value: unknown) => typeof value === "string" && /^[0-9a-f]{64}$/.test(value);
const bounded = (value: unknown, max: number): value is string => typeof value === "string" && !!value.trim() && Array.from(value).length <= max && !/[\u0000-\u001f\u007f]/.test(value);

export function knownPackingSelection(value: unknown): EvidencePackingSelection | null {
  if (!keys(value, ["version", "position", "chunk_id", "source_group_id", "source_snapshot_sha256", "diversity_group_id", "source_group_basis", "group_basis", "role_hint", "selection_reason", "scientific_acceptance"])
    || value.version !== "evidence-pack-item/1.0.0" || !integer(value.position, 1, 20) || !bounded(value.chunk_id, 200)
    || typeof value.source_group_id !== "string" || !/^src:[0-9a-f]{64}$/.test(value.source_group_id)
    || typeof value.diversity_group_id !== "string" || !/^div:[0-9a-f]{64}$/.test(value.diversity_group_id)
    || !oneOf(value.source_group_basis, ["source_snapshot", "legacy_paper"])
    || !oneOf(value.group_basis, ["accepted_work_mapping", "source_snapshot", "legacy_paper"])
    || !oneOf(value.role_hint, ["methods", "results", "table", "other"])
    || !oneOf(value.selection_reason, ["source_diversity", "source_coverage", "complementary_role"])
    || value.scientific_acceptance !== false) return null;
  if (value.source_group_basis === "source_snapshot" ? !hash(value.source_snapshot_sha256) : value.source_snapshot_sha256 !== null) return null;
  if (value.group_basis !== "accepted_work_mapping" && value.group_basis !== value.source_group_basis
    || value.selection_reason === "complementary_role" && (value.source_group_basis === "legacy_paper" || value.role_hint === "other")) return null;
  return value as unknown as EvidencePackingSelection;
}

export function knownInputBudget(value: unknown): RagInputBudgetReport | null {
  if (!keys(value, ["status", "model", "profile", "request_sha256", "payload_bytes", "byte_limit", "input_tokens", "max_input_tokens", "count_method", "generation_started", "scientific_acceptance"])
    || !oneOf(value.status, ["not_requested", "counted", "rejected", "unavailable"])
    || value.model !== null && (typeof value.model !== "string" || !/^gemini-[A-Za-z0-9][A-Za-z0-9._-]{0,119}$/.test(value.model))
    || value.profile !== "sclib-gemini-text-rag/1.0.0" || value.request_sha256 !== null && !hash(value.request_sha256)
    || !nullableInteger(value.payload_bytes, 1, Number.MAX_SAFE_INTEGER) || !integer(value.byte_limit, 1, 1048576)
    || !nullableInteger(value.input_tokens, 1, 2147483647) || !nullableInteger(value.max_input_tokens, 1, 131072)
    || value.count_method !== "provider_count_tokens" || value.generation_started !== null && typeof value.generation_started !== "boolean"
    || value.scientific_acceptance !== false) return null;
  if (value.status === "not_requested" && [value.model, value.request_sha256, value.payload_bytes,
    value.input_tokens, value.max_input_tokens, value.generation_started].some(field => field !== null)
    || value.status === "counted" && value.generation_started !== true
    || value.status === "rejected" && value.generation_started) return null;
  if ((value.status === "counted" || value.generation_started)
    && (value.model === null || value.request_sha256 === null || !integer(value.payload_bytes, 1, value.byte_limit)
      || !integer(value.input_tokens, 1, 2147483647) || !integer(value.max_input_tokens, 1, 131072)
      || value.input_tokens > value.max_input_tokens)) return null;
  return value as unknown as RagInputBudgetReport;
}

export type PackingSource = { index?: unknown; paper_id?: unknown; packing_info?: unknown; evidence_provenance?: unknown };

/** Validate declared source-snapshot grouping; this cannot establish a root or Work mapping. */
export function knownPackingInventory(sources: PackingSource[]): EvidencePackingSelection[] | null {
  if (sources.length > 20) return null;
  const items: EvidencePackingSelection[] = [];
  const groups = new Map<string, { paper: string; snapshot: string | null; diversity: string; basis: string; count: number }>();
  const sourceKeys = new Map<string, string>(), diversityBases = new Map<string, string>();
  const chunks = new Set<string>(), diversity = new Set<string>();
  for (const [position, source] of sources.entries()) {
    const item = knownPackingSelection(source.packing_info);
    if (!item || !bounded(source.paper_id, 100) || source.index !== position + 1 || item.position !== source.index || chunks.has(item.chunk_id)) return null;
    const existing = groups.get(item.source_group_id);
    const expectedReason = !diversity.has(item.diversity_group_id) ? "source_diversity" : !existing ? "source_coverage" : "complementary_role";
    if (item.selection_reason !== expectedReason) return null;
    if (item.selection_reason === "complementary_role" && (knownEvidenceProvenance(source.evidence_provenance)?.chunk_kind !== "original_passage"
      || items.some(other => other.source_group_id === item.source_group_id && other.role_hint === item.role_hint))) return null;
    if (existing && (existing.paper !== source.paper_id || existing.snapshot !== item.source_snapshot_sha256
      || existing.diversity !== item.diversity_group_id || existing.basis !== item.group_basis)) return null;
    const sourceKey = JSON.stringify([source.paper_id, item.source_snapshot_sha256]);
    if (sourceKeys.has(sourceKey) && sourceKeys.get(sourceKey) !== item.source_group_id
      || diversityBases.has(item.diversity_group_id) && diversityBases.get(item.diversity_group_id) !== item.group_basis) return null;
    const count = (existing?.count ?? 0) + 1;
    if (count > (item.source_group_basis === "legacy_paper" ? 1 : 3)) return null;
    groups.set(item.source_group_id, { paper: source.paper_id, snapshot: item.source_snapshot_sha256, diversity: item.diversity_group_id, basis: item.group_basis, count });
    sourceKeys.set(sourceKey, item.source_group_id);
    diversityBases.set(item.diversity_group_id, item.group_basis);
    diversity.add(item.diversity_group_id);
    chunks.add(item.chunk_id);
    items.push(item);
  }
  for (const item of items) {
    if (item.group_basis !== "accepted_work_mapping" && items.some(other => other.diversity_group_id === item.diversity_group_id && other.source_group_id !== item.source_group_id)
      || item.group_basis === "accepted_work_mapping" && items.filter(other => other.diversity_group_id === item.diversity_group_id).length > 3) return null;
  }
  return items;
}

const exclusions = ["payload_budget", "base_payload_budget", "chunk_limit", "source_limit", "work_limit", "duplicate_content", "role_already_represented", "not_complementary_original"];
const reasons = ["packing_not_requested", "no_admitted_candidates", "base_payload_budget_exceeded", "payload_budget_excluded", "selection_limits_applied", "duplicate_content_removed", "complementarity_not_established", "packing_unavailable", "selected_context_withheld"];

export function knownPackingSummary(value: unknown, sources: PackingSource[]): EvidencePackingSummary | null {
  if (!keys(value, ["version", "status", "candidate_count", "selected_count", "source_group_count", "diversity_group_count", "payload_bytes", "byte_budget", "byte_count_method", "max_chunks", "max_per_source", "max_per_work", "reason_counts", "reason_codes", "independent_support_count", "independence_status", "scientific_acceptance"])
    || value.version !== "evidence-packing/1.0.0" || !oneOf(value.status, ["not_requested", "packed", "empty", "base_budget_exceeded", "withheld", "unavailable"])
    || !integer(value.candidate_count, 0, 300) || !integer(value.selected_count, 0, 20)
    || !integer(value.source_group_count, 0, 20) || !integer(value.diversity_group_count, 0, 20)
    || !nullableInteger(value.payload_bytes, 0, 67108864) || !nullableInteger(value.byte_budget, 1, 16777216)
    || value.byte_count_method !== "utf8-full-payload/1" || !integer(value.max_chunks, 1, 20)
    || !integer(value.max_per_source, 1, 3) || !integer(value.max_per_work, 1, 3)
    || !row(value.reason_counts) || Object.entries(value.reason_counts).some(([key, count]) => !exclusions.includes(key) || !integer(count, 1, 300))
    || !Array.isArray(value.reason_codes) || value.reason_codes.length > 9 || value.reason_codes.some(item => !oneOf(item, reasons))
    || new Set(value.reason_codes).size !== value.reason_codes.length || value.independent_support_count !== null
    || value.independence_status !== "independence_not_established" || value.scientific_acceptance !== false) return null;
  if (!(value.diversity_group_count <= value.source_group_count && value.source_group_count <= value.selected_count
    && value.selected_count <= value.candidate_count && value.selected_count <= value.max_chunks)) return null;
  if (["not_requested", "withheld", "unavailable"].includes(value.status as string)) {
    if (value.selected_count !== 0 || value.source_group_count !== 0 || value.diversity_group_count !== 0
      || value.payload_bytes !== null || Object.keys(value.reason_counts).length) return null;
    if (value.status === "not_requested" ? value.candidate_count !== 0 || value.byte_budget !== null
      || sources.some(source => source.packing_info !== null && source.packing_info !== undefined) : sources.length > 0) return null;
  } else {
    if (!integer(value.payload_bytes, 0, 67108864) || !integer(value.byte_budget, 1, 16777216)
      || Object.values(value.reason_counts).reduce<number>((total, count) => total + (count as number), 0) !== value.candidate_count - value.selected_count) return null;
    if (value.status === "base_budget_exceeded" ? value.selected_count !== 0 || value.payload_bytes <= value.byte_budget
      : value.payload_bytes > value.byte_budget || (value.selected_count > 0) !== (value.status === "packed")) return null;
    const items = knownPackingInventory(sources);
    if (!items || items.length !== value.selected_count || new Set(items.map(item => item.source_group_id)).size !== value.source_group_count
      || new Set(items.map(item => item.diversity_group_id)).size !== value.diversity_group_count
      || items.some(item => items.filter(other => other.source_group_id === item.source_group_id).length > (value.max_per_source as number)
        || item.group_basis === "accepted_work_mapping" && items.filter(other => other.diversity_group_id === item.diversity_group_id).length > (value.max_per_work as number))) return null;
  }
  return value as unknown as EvidencePackingSummary;
}
